source/taskmanager/src/main/java/nl/aerius/taskmanager/TaskDispatcher.java

```diff
@@ -43,26 +43,40 @@
     WAIT_FOR_WORKER, WAIT_FOR_TASK, DISPATCH_TASK;
   }
 
-  private final ConcurrentHashMap<TaskConsumer, Semaphore> taskConsumerLocks = new ConcurrentHashMap<>();
+  private final Map<TaskConsumer, Semaphore> taskConsumerLocks = new ConcurrentHashMap<>();
   private final WorkerPool workerPool;
-  private final TaskScheduler scheduler;
+  private final TaskScheduler<?> scheduler;
 
   private boolean running;
   private State state;
   private final String workerQueueName;
 
-  public TaskDispatcher(final String workerQueueName, final TaskScheduler scheduler, final WorkerPool workerPool) {
+  public TaskDispatcher(final String workerQueueName, final TaskScheduler<?> scheduler, final WorkerPool workerPool) {
     this.workerQueueName = workerQueueName;
     this.scheduler = scheduler;
     this.workerPool = workerPool;
   }
 
+  /**
+   * Forwards task to the scheduler. It allows only one task per taskconsumer to be ready to be scheduled.
+   * To do this it blocks completing this method after passing the task to the scheduler.
+   * Since each taskconsumer runs in it's own thread this blocks per taskconsumer.
+   */
   @Override
   public void forwardTask(final Task task) {
     final TaskConsumer taskConsumer = task.getTaskConsumer();
-    lockClient(taskConsumer);
     scheduler.addTask(task);
     LOG.debug("Task for {} added to scheduler {}", workerQueueName, task.getId());
+    lockClient(taskConsumer);
+  }
+
+  @Override
+  public void killTasks() {
+    // Remove all locks to let RabbitMQ message handlers continue and close shutdown.
+    taskConsumerLocks.forEach((t, s) -> s.release());
+    // Tell the scheduler to mark all tasks it tracks to be dead.
+    scheduler.killTasks();
```

Comment:
Doesn't this need to be switched to avoid a race condition where scheduler kills a task that has just been added due to releasing the locks?